### PR TITLE
Add specific support for the .NET 9 TFM

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
For .NET 9 TFM, use the COM source generator
which supports cross-platform COM interop via
the [`ComWrappers`](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.comwrappers) API.

### CODE REVIEW

File: src/Microsoft.DiaSymReader/Metadata/IMetadataImport.cs [hunk 1]
```diff
@@ -26,37 +33,37 @@ internal unsafe interface IMetadataImport
         [PreserveSig]
         int EnumTypeDefs(
             ref void* enumHandle,
-            [Out]int* typeDefs, 
+            int* typeDefs,
```
Comment (tmat) [inline src/Microsoft.DiaSymReader/Metadata/IMetadataImport.cs:36]: Are the Ref/In/Out attributes not needed for netstandard2.0?

File: src/Microsoft.DiaSymReader/Reader/ISymUnmanagedBinder2.cs [hunk 2]
```diff
@@ -7,12 +7,13 @@
...
 {
-    [ComImport]
     [Guid("ACCEE350-89AF-4ccb-8B40-1C2C4C6F9434")]
     [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
     [ComVisible(false)]
-    public interface ISymUnmanagedBinder2 : ISymUnmanagedBinder
+    [GeneratedWhenPossibleComInterface]
+    public partial interface ISymUnmanagedBinder2 : ISymUnmanagedBinder
     {
+#if NETSTANDARD2_0
```
Comment (tmat) [inline src/Microsoft.DiaSymReader/Reader/ISymUnmanagedBinder2.cs:16]: Are there any issues with these methods in NET9 build?

File: src/Microsoft.DiaSymReader/Reader/ISymUnmanagedMethod2.cs [hunk 3]
```diff
@@ -7,12 +7,13 @@
...
 {
-    [ComImport]
     [Guid("5da320c8-9c2c-4e5a-b823-027e0677b359")]
     [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
     [ComVisible(false)]
-    public interface ISymUnmanagedMethod2 : ISymUnmanagedMethod
+    [GeneratedWhenPossibleComInterface]
+    public partial interface ISymUnmanagedMethod2 : ISymUnmanagedMethod
     {
+#if NETSTANDARD2_0
```
Comment (tmat) [inline src/Microsoft.DiaSymReader/Reader/ISymUnmanagedMethod2.cs:16]: Why omit these from NET9?

File: src/Microsoft.DiaSymReader/Reader/ISymUnmanagedReader5.cs [hunk 4]
```diff
@@ -4,16 +4,20 @@
...
 {
-    [ComImport]
     [Guid("6576c987-7e8d-4298-a6e1-6f9783165f07")]
     [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
     [ComVisible(false)]
-    public interface ISymUnmanagedReader5 : ISymUnmanagedReader4
+    [GeneratedWhenPossibleComInterface]
+    public unsafe partial interface ISymUnmanagedReader5 : ISymUnmanagedReader4
     {
+#if NETSTANDARD2_0
```
Comment (tmat) [inline src/Microsoft.DiaSymReader/Reader/ISymUnmanagedReader5.cs:20]: Same question

File: src/Microsoft.DiaSymReader/Metadata/IMetadataImport.cs (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DiaSymReader/Metadata/IMetadataImport.cs:36]: The attributes are rarely needed. In this case, the types are pointers so there is literally no impact.

File: src/Microsoft.DiaSymReader/SymUnmanagedFactory.cs [hunk 5]
```diff
@@ -41,38 +53,51 @@ internal static string DiaSymReaderModuleName
 
         [DefaultDllImportSearchPaths(DllImportSearchPath.AssemblyDirectory | DllImportSearchPath.SafeDirectories)]
         [DllImport(DiaSymReaderModuleName32, EntryPoint = CreateSymReaderFactoryName)]
-        private static extern void CreateSymReader32(ref Guid id, [MarshalAs(UnmanagedType.IUnknown)] out object symReader);
+        private static unsafe extern void CreateSymReader32([MarshalAs(UnmanagedType.LPStruct)] Guid id, IntPtr* symReader);
```
Comment (tmat) [inline src/Microsoft.DiaSymReader/SymUnmanagedFactory.cs:56]: Changing `object` to `IntPtr*` seems like a breaking change.

File: src/Microsoft.DiaSymReader/Reader/ISymUnmanagedBinder2.cs (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DiaSymReader/Reader/ISymUnmanagedBinder2.cs:16]: Historically COM interop requires redefinition of inherited methods so the vtables are properly created. For the COM source generator in .NET 8+, the system respects "inheritance" from other COM interfaces. This means there is no need to redefine them.

File: src/Microsoft.DiaSymReader/SymUnmanagedFactory.cs [hunk 6]
```diff
@@ -41,38 +53,51 @@ internal static string DiaSymReaderModuleName
...
-        private static extern void CreateSymWriterArm64(ref Guid id, [MarshalAs(UnmanagedType.IUnknown)] out object symWriter);
+        private static unsafe extern void CreateSymWriterArm64([MarshalAs(UnmanagedType.LPStruct)] Guid id, IntPtr* symWriter);
 
-        [DllImport("kernel32")]
-        private static extern IntPtr LoadLibrary(string path);
 
-        [DllImport("kernel32")]
+#if NETSTANDARD2_0
+        [DllImport("kernel32", CharSet = CharSet.Unicode, ExactSpelling = true)]
+        private static extern IntPtr LoadLibraryW(string path);
```
Comment (tmat) [inline src/Microsoft.DiaSymReader/SymUnmanagedFactory.cs:81]: You can only #if-def the attribute, not the signature, which is the same.

File: src/Microsoft.DiaSymReader/SymUnmanagedFactory.cs (hunk 5 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DiaSymReader/SymUnmanagedFactory.cs:56]: Yes, but it is a `private` method so not exposed.

File: src/Microsoft.DiaSymReader/SymUnmanagedFactory.cs (hunk 6 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DiaSymReader/SymUnmanagedFactory.cs:81]: They are not the same. One is `extern` the other is `partial`.

File: src/Microsoft.DiaSymReader/Reader/ISymUnmanagedMethod2.cs (hunk 3 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DiaSymReader/Reader/ISymUnmanagedMethod2.cs:16]: See https://github.com/dotnet/symreader/pull/320#discussion_r1637048471

File: src/Microsoft.DiaSymReader/Reader/ISymUnmanagedReader5.cs (hunk 4 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DiaSymReader/Reader/ISymUnmanagedReader5.cs:20]: See https://github.com/dotnet/symreader/pull/320#discussion_r1637048471

File: src/Microsoft.DiaSymReader/Reader/ISymUnmanagedBinder2.cs (hunk 2 above)
Comment (tmat) [reply src/Microsoft.DiaSymReader/Reader/ISymUnmanagedBinder2.cs:16]: I see, could you please add a comment to the code?

File: src/Microsoft.DiaSymReader/SymUnmanagedFactory.cs (hunk 5 above)
Comment (tmat) [reply src/Microsoft.DiaSymReader/SymUnmanagedFactory.cs:56]: Ah, missed that. Thx.

File: src/Microsoft.DiaSymReader/Reader/ISymUnmanagedBinder2.cs (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/Microsoft.DiaSymReader/Reader/ISymUnmanagedBinder2.cs:16]: Sure. I assume in all the places where they are commented out?

**Review by tmat — APPROVED** [review]

**Pull request merged**